crates/daphne/src/roles/leader/mod.rs

```diff
@@ -223,6 +223,47 @@ pub async fn handle_upload_req<A: DapLeader>(
         .into());
     }
 
+    match (
+        &report.report_metadata.public_extensions,
+        task_config.version,
+    ) {
+        (Some(extensions), DapVersion::Latest) => {
+            let mut unknown_extensions = Vec::<u16>::new();
+            if crate::protocol::no_duplicates(extensions.iter()).is_err() {
+                return Err(DapError::Abort(DapAbort::InvalidMessage {
+                    detail: "Repeated public extension".into(),
+                    task_id,
+                }));
+            };
+            for extension in extensions {
+                match extension {
+                    Extension::Taskprov => (),
+                    Extension::NotImplemented { typ, .. } => unknown_extensions.push(*typ),
+                }
+            }
+
+            if !unknown_extensions.is_empty() {
+                return match DapAbort::unsupported_extension(&task_id, &unknown_extensions) {
+                    Ok(abort) => Err::<(), DapError>(abort.into()),
+                    Err(fatal) => Err(fatal),
+                };
+            }
+        }
+        (None, DapVersion::Draft09) => (),
+        (Some(_), DapVersion::Draft09) => {
+            return Err(DapError::Abort(DapAbort::version_mismatch(
+                DapVersion::Draft09,
+                DapVersion::Latest,
+            )))
+        }
+        (None, DapVersion::Latest) => {
+            return Err(DapError::Abort(DapAbort::version_mismatch(
+                DapVersion::Latest,
+                DapVersion::Draft09,
+            )))
+        }
```

Comment:
I'm not sure either of these arms could ever be reached, assuming the code that sets `public_extensions` is correct. If I were you, I would use the `public_extension.is_some()` to deduce which version was negotiated. Tnen we could replace this big ugly `match` with:

```rust
if let Some(public_extensions) = &report.report_metadata.public_extensions {
    ...  // check for duplicates, unorecgonized values, and so on
}
```